File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/types/digitaltwin/PropertyElementDescriptor.java
package org.opcfoundation.webserver.types.digitaltwin;

import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.opcfoundation.webserver.addressspace.nodes.UaVariableType;
import org.springframework.lang.Nullable;

public class PropertyElementDescriptor {
    private final String id;
    private final LocalizedText displayName;
    private final NodeId typeId;

    public PropertyElementDescriptor(
            String propertyId,
            LocalizedText displayName,
            @Nullable UaVariableType variableType)
    {
        this.id = propertyId;
        this.displayName = displayName;
        this.typeId = (null != variableType) ? variableType.nodeId() : NodeIds.BaseDataVariableType;
    }

    public String getId() {
        return id;
    }

    public LocalizedText getDisplayName() {
        return displayName;
    }

    public NodeId getTypeId() {
        return typeId;
    }
}

File: opcua-webservice-node/src/server/types/digitaltwin/PropertyElementDescriptor.ts
import { UaLocalizedText, UaNodeId, VariableTypeIds } from "opcua-webapi-ts";
import { UaVariableType } from "../../addressspace";

export class PropertyElementDescriptor {
    private readonly _id: string;
    private readonly _displayName: UaLocalizedText;
    private readonly _typeId: UaNodeId;

    constructor(propertyId: string, displayName: UaLocalizedText, variableType?: UaVariableType) {
        this._id = propertyId;
        this._displayName = displayName;
        this._typeId = (variableType) ? variableType.nodeId
            : UaNodeId.from(VariableTypeIds.BaseDataVariableType);
    }

    get id(): string {
        return this._id;
    }

    get displayName(): UaLocalizedText {
        return this._displayName;
    }

    get typeId(): UaNodeId {
        return this._typeId;
    }
}